core/inherents/src/lib.rs

```diff
@@ -0,0 +1,527 @@
+			.and_then(|(i, v)| {
+				if i.len() == v.len() {
+					Some(Self {
+						data: i.into_iter().zip(v.into_iter()).collect()
+					})
+				} else {
+					None
+				}
+			})
+	}
+}
+
+/// The result of checking inherents.
+///
+/// It either returns okay for all checks, or stores all occurred errors.
+#[derive(Encode, Decode)]
+pub struct CheckInherentsResult {
+	/// Did the check succeeded?
+	okay: bool,
+	/// Did we encountered a fatal error?
+	fatal_error: bool,
+	/// We use the `InherentData` to store our errors.
+	errors: InherentData,
+}
+
+impl CheckInherentsResult {
+	/// Create a new instance.
+	pub fn new() -> Self {
+		Self {
+			okay: true,
+			errors: InherentData::new(),
+			fatal_error: false,
+		}
+	}
+
+	/// Put an error into the result.
+	///
+	/// This makes this result resolve to `ok() == false`.
+	///
+	/// # Parameters
+	///
+	/// - identifier - The identifier of the inherent that generated the error.
+	/// - error - The error that will be encoded.
+	pub fn put_error<E: codec::Encode + IsFatalError>(
+		&mut self, identifier: InherentIdentifier, error: &E
+	) -> Result<(), RuntimeString> {
+		// Don't accept any other error
+		if self.fatal_error {
+			return Err("No other errors are accepted after an hard error!".into())
+		}
+
+		if error.is_fatal_error() {
+			// remove the other errors.
+			self.errors.data.clear();
+		}
+
+		self.errors.put_data(identifier, error)?;
+
+		self.okay = false;
+		self.fatal_error = error.is_fatal_error();
+		Ok(())
+	}
+
+	/// Get an error out of the result.
+	///
+	/// # Return
+	///
+	/// - `Ok(Some(I))` if the error could be found and deserialized.
+	/// - `Ok(None)` if the error could not be found.
+	/// - `Err(_)` if the error could be found, but deserialization did not work.
+	pub fn get_error<E: codec::Decode>(
+		&self, identifier: &InherentIdentifier
+	) -> Result<Option<E>, RuntimeString> {
+		self.errors.get_data(identifier)
+	}
+
+	/// Convert into an iterator over all contained errors.
+	pub fn into_errors(self) -> IntoIter<InherentIdentifier, Vec<u8>> {
+		self.errors.data.into_iter()
+	}
+
+	/// Is this result ok?
+	pub fn ok(&self) -> bool {
+		self.okay
+	}
+
+	/// Is this a fatal error?
+	pub fn fatal_error(&self) -> bool {
+		self.fatal_error
+	}
+}
+
+/// All `InherentData` providers.
+#[cfg(feature = "std")]
+#[derive(Clone)]
+pub struct InherentDataProviders {
+	providers: Arc<RwLock<Vec<Box<dyn ProvideInherentData + Send + Sync>>>>,
+}
+
+#[cfg(feature = "std")]
+impl InherentDataProviders {
+	/// Create a new instance.
+	pub fn new() -> Self {
+		Self {
+			providers: Default::default(),
+		}
+	}
+
+	/// Register an `InherentData` provider.
+	///
+	/// The registration order is preserved and this order will also be used when creating the
+	/// inherent data.
+	///
+	/// # Result
+	///
+	/// Will return an error, if a provider with the same identifier already exists.
+	pub fn register_provider<P: ProvideInherentData + Send + Sync +'static>(
```

Comment:
for swappable consensus, we would probably want to register and maybe deregister when we change consensus algorithms. Since each provider carries around an implicit "dependency tree" of other providers, would deregistration be easy? or is it probably fine to never deregister (assuming any extra bits of inherent data will be ignored)?

Reply:
We could provide a mechanism to unregister a provider, but I would postpone this to a later pull request. All inherent data that is not used, is just ignored. So, there will be no error when we do not use something.
We would also need to make sure that we do not unregister stuff, that is required by another provider. So, we would probably start to track some dependencies between the providers.